#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <QScrollBar>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    // ========================================================
    // –ù–ê–õ–ê–®–¢–£–í–ê–ù–ù–Ø –í–Ü–ó–£–ê–õ–¨–ù–û–ì–û –°–¢–ò–õ–Æ (–¢–ï–ú–ù–ê –¢–ï–ú–ê)
    // ========================================================

    // –í—Å—Ç–∞–Ω–æ–≤–ª—é—î–º–æ –∑–∞–≥–∞–ª—å–Ω–∏–π —Å—Ç–∏–ª—å –¥–ª—è –≤—Å—å–æ–≥–æ –≤—ñ–∫–Ω–∞
    this->setStyleSheet(
        // –¢–µ–º–Ω–∏–π —Ñ–æ–Ω –≤—ñ–∫–Ω–∞
        "QMainWindow { background-color: #2b2b2b; }"

        // –°—Ç–∏–ª—å –¥–ª—è –ø–æ–ª—è –ª–æ–≥—É (—Å—Ö–æ–∂–∏–π –Ω–∞ —Å—Ç–∞—Ä–∏–π –ø–µ—Ä–≥–∞–º–µ–Ω—Ç –∞–±–æ —Ç–µ—Ä–º—ñ–Ω–∞–ª)
        "QTextBrowser#gameLog {"
        "   background-color: #1e1e1e;"
        "   color: #e0e0e0;"
        "   border: 2px solid #5c5c5c;"
        "   border-radius: 5px;"
        "   font-family: 'Consolas', 'Courier New', monospace;"
        "   font-size: 14px;"
        "   padding: 10px;"
        "}"

        // –°—Ç–∏–ª—å –¥–ª—è –∫–Ω–æ–ø–æ–∫ (–∑–∞–≥–∞–ª—å–Ω–∏–π)
        "QPushButton {"
        "   background-color: #4a4a4a;"
        "   color: white;"
        "   border: 1px solid #3a3a3a;"
        "   border-radius: 4px;"
        "   padding: 8px;"
        "   font-weight: bold;"
        "   font-size: 12px;"
        "}"
        "QPushButton:hover { background-color: #5a5a5a; }" // –ü—Ä–∏ –Ω–∞–≤–µ–¥–µ–Ω–Ω—ñ —Å–≤—ñ—Ç–ª—ñ—à–∞—î
        "QPushButton:pressed { background-color: #3a3a3a; }" // –ü—Ä–∏ –Ω–∞—Ç–∏—Å–∫–∞–Ω–Ω—ñ —Ç–µ–º–Ω—ñ—î
        "QPushButton:disabled { background-color: #333333; color: #777777; }" // –ù–µ–∞–∫—Ç–∏–≤–Ω–∞ –∫–Ω–æ–ø–∫–∞

        // –°–ø–µ—Ü—ñ–∞–ª—å–Ω–∏–π —Å—Ç–∏–ª—å –¥–ª—è –∫–Ω–æ–ø–∫–∏ –ê–¢–ê–ö–ò (—á–µ—Ä–≤–æ–Ω–∞)
        "QPushButton#btnAttack {"
        "   background-color: #c0392b;"
        "   border: 1px solid #a93226;"
        "}"
        "QPushButton#btnAttack:hover { background-color: #e74c3c; }"

        // –°–ø–µ—Ü—ñ–∞–ª—å–Ω–∏–π —Å—Ç–∏–ª—å –¥–ª—è –∫–Ω–æ–ø–∫–∏ –ù–û–í–ê –ì–†–ê (–∑–µ–ª–µ–Ω–∞)
        "QPushButton#btnStart {"
        "   background-color: #27ae60;"
        "}"
        "QPushButton#btnStart:hover { background-color: #2ecc71; }"

        // –°—Ç–∏–ª—å –¥–ª—è —Å–º—É–∂–∫–∏ –∑–¥–æ—Ä–æ–≤'—è
        "QProgressBar {"
        "   border: 2px solid #5c5c5c;"
        "   border-radius: 5px;"
        "   text-align: center;"
        "   color: white;"
        "   background-color: #1e1e1e;"
        "}"
        );

    game = new Game(this);

    // --- 1. –°–ò–ì–ù–ê–õ–ò –í–Ü–î –ì–†–ò ---

    // –õ–æ–≥—É–≤–∞–Ω–Ω—è —Ç–µ–∫—Å—Ç—É
    connect(game, &Game::logMessage, this, [this](QString msg){
        ui->gameLog->append(msg);
        ui->gameLog->verticalScrollBar()->setValue(ui->gameLog->verticalScrollBar()->maximum());
    });

    // –û–Ω–æ–≤–ª–µ–Ω–Ω—è –∫–Ω–æ–ø–æ–∫ —ñ HP, –∫–æ–ª–∏ —â–æ—Å—å –∑–º—ñ–Ω—é—î—Ç—å—Å—è –≤ –≥—Ä—ñ
    connect(game, &Game::statsUpdated, this, &MainWindow::updateUI);
    connect(game, &Game::roomUpdated, this, &MainWindow::updateUI);
    connect(game, &Game::gameStarted, this, &MainWindow::updateUI);

    // –û–±—Ä–æ–±–∫–∞ –∫—ñ–Ω—Ü—è –≥—Ä–∏
    connect(game, &Game::gameOver, this, [this](bool victory){
        updateUI(); // –¶–µ –≤–∏–∫–ª–∏—á–µ –Ω–∞—à—É –Ω–æ–≤—É –ª–æ–≥—ñ–∫—É, —è–∫–∞ —Å—Ö–æ–≤–∞—î –∑–∞–π–≤—ñ –∫–Ω–æ–ø–∫–∏

        if(victory) {
            ui->gameLog->append("\nüèÜ –í–Ü–¢–ê–Ñ–ú–û! –í–ò –í–ò–ì–†–ê–õ–ò! (–í—Å—ñ –≤–æ—Ä–æ–≥–∏ –∑–Ω–∏—â–µ–Ω—ñ)");
            // –ü—Ä–∏ –ø–µ—Ä–µ–º–æ–∑—ñ —Ç–µ–∂ —Ö–æ–≤–∞—î–º–æ –≤—Å–µ –∑–∞–π–≤–µ
            ui->btnAttack->setVisible(false);
            ui->btnMove1->setVisible(false);
            ui->btnMove2->setVisible(false);
            ui->btnStart->setVisible(true);
        }
        else {
            ui->gameLog->append("\nüíÄ –ì–†–ê –ó–ê–ö–Ü–ù–ß–ï–ù–ê. –°–ø—Ä–æ–±—É–π—Ç–µ —â–µ —Ä–∞–∑!");
        }
    });

    // --- 2. –ö–ù–û–ü–ö–ò (UI -> –ì–†–ê) ---

    // –ù–æ–≤–∞ –≥—Ä–∞
    connect(ui->btnStart, &QPushButton::clicked, this, [this](){
        ui->gameLog->clear();
        ui->hpBar->reset(); // –°–∫–∏–¥–∞–Ω–Ω—è –∫–æ–ª—å–æ—Ä—É
        // –í–º–∏–∫–∞—î–º–æ –∫–Ω–æ–ø–∫–∏ –Ω–∞–∑–∞–¥
        ui->btnMove1->setEnabled(true);
        ui->btnMove2->setEnabled(true);

        game->startNewGame("–ì–µ—Ä–æ–π", 0); // 0 = –í–æ—ó–Ω
    });

    // –ê—Ç–∞–∫–∞
    connect(ui->btnAttack, &QPushButton::clicked, game, &Game::actionAttack);

    // –†—É—Ö (–ö–Ω–æ–ø–∫–∞ 1)
    connect(ui->btnMove1, &QPushButton::clicked, this, [this](){
        // –Ø–∫—â–æ –∫–Ω–æ–ø–∫–∞ –∞–∫—Ç–∏–≤–Ω–∞, –∑–Ω–∞—á–∏—Ç—å –≤–∏—Ö—ñ–¥ 0 —ñ—Å–Ω—É—î
        game->actionMove(0);
    });

    // –†—É—Ö (–ö–Ω–æ–ø–∫–∞ 2)
    connect(ui->btnMove2, &QPushButton::clicked, this, [this](){
        game->actionMove(1);
    });

    // –ü–æ—á–∞—Ç–∫–æ–≤–∏–π —Å—Ç–∞–Ω: —Ö–æ–≤–∞—î–º–æ –∫–Ω–æ–ø–∫–∏ –±–æ—é —ñ —Ä—É—Ö—É –¥–æ —Å—Ç–∞—Ä—Ç—É –≥—Ä–∏
    ui->btnAttack->setVisible(false);
    ui->btnMove1->setVisible(false);
    ui->btnMove2->setVisible(false);
    ui->hpBar->setValue(0);
}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::updateUI()
{
    // 1. –û–Ω–æ–≤–ª–µ–Ω–Ω—è HP (—Ü–µ —Ä–æ–±–∏–º–æ –∑–∞–≤–∂–¥–∏, –Ω–∞–≤—ñ—Ç—å —è–∫—â–æ –º–µ—Ä—Ç–≤–∏–π)
    int hp = game->getPlayerHP();
    int maxHp = game->getPlayerMaxHP();
    ui->hpBar->setMaximum(maxHp);
    ui->hpBar->setValue(hp);
    ui->hpBar->setFormat("%v / %m HP");

    // –ö–æ–ª—å–æ—Ä–∏ —Å–º—É–∂–∫–∏
    if (hp > maxHp * 0.5) ui->hpBar->setStyleSheet("QProgressBar::chunk { background-color: #2ecc71; }");
    else if (hp > maxHp * 0.25) ui->hpBar->setStyleSheet("QProgressBar::chunk { background-color: #f1c40f; }");
    else ui->hpBar->setStyleSheet("QProgressBar::chunk { background-color: #e74c3c; }");

    // --- –í–ê–ñ–õ–ò–í–ê –ó–ú–Ü–ù–ê: –Ø–ö–©–û –ì–†–ê–í–ï–¶–¨ –ú–ï–†–¢–í–ò–ô - –•–û–í–ê–Ñ–ú–û –í–°–ï –Ü –í–ò–•–û–î–ò–ú–û ---
    if (hp <= 0) {
        ui->btnAttack->setVisible(false);
        ui->btnMove1->setVisible(false);
        ui->btnMove2->setVisible(false);
        ui->btnStart->setVisible(true); // –ö–Ω–æ–ø–∫—É "–ù–æ–≤–∞ –≥—Ä–∞" –∑–∞–≤–∂–¥–∏ –ø–æ–∫–∞–∑—É—î–º–æ –ø—Ä–∏ —Å–º–µ—Ä—Ç—ñ
        return;
    }
    // -------------------------------------------------------------------

    // 2. –õ–æ–≥—ñ–∫–∞ –∫–Ω–æ–ø–∫–∏ –ê–¢–ê–ö–ò
    if (game->getEnemyHP() > 0) {
        ui->btnAttack->setVisible(true);
        ui->btnAttack->setEnabled(true);
        ui->btnAttack->setText(QString("–ê–¢–ê–ö–ê (%1 HP)").arg(game->getEnemyHP()));
    } else {
        ui->btnAttack->setVisible(false);
    }

    // 3. –õ–æ–≥—ñ–∫–∞ –∫–Ω–æ–ø–æ–∫ –†–£–•–£
    // (–ú–∏ —Å—é–¥–∏ –¥—ñ–π–¥–µ–º–æ —Ç—ñ–ª—å–∫–∏ —è–∫—â–æ –≥—Ä–∞–≤–µ—Ü—å –∂–∏–≤–∏–π, –±–æ –≤–∏—â–µ —Å—Ç–æ—ó—Ç—å return)

    // –Ø–∫—â–æ —î –∂–∏–≤–∏–π –≤–æ—Ä–æ–≥ - —Ä—É—Ö –∑–∞–±–æ—Ä–æ–Ω–µ–Ω–æ
    if (game->getEnemyHP() > 0) {
        ui->btnMove1->setVisible(false);
        ui->btnMove2->setVisible(false);
    }
    else {
        // –Ø–∫—â–æ –≤–æ—Ä–æ–≥—ñ–≤ –Ω–µ–º–∞—î - –ø–æ–∫–∞–∑—É—î–º–æ –≤–∏—Ö–æ–¥–∏
        QVector<QString> exits = game->getAvailableExits();

        if (exits.size() > 0) {
            ui->btnMove1->setVisible(true);
            ui->btnMove1->setText(exits[0]);
        } else {
            ui->btnMove1->setVisible(false);
        }

        if (exits.size() > 1) {
            ui->btnMove2->setVisible(true);
            ui->btnMove2->setText(exits[1]);
        } else {
            ui->btnMove2->setVisible(false);
        }
    }
}
